pkg/providers/nutanix/validator.go

```diff
@@ -0,0 +1,195 @@
+package nutanix
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	prismgoclient "github.com/nutanix-cloud-native/prism-go-client"
+	"github.com/nutanix-cloud-native/prism-go-client/utils"
+	"github.com/nutanix-cloud-native/prism-go-client/v3"
+	"go.uber.org/multierr"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+)
+
+// Validator is a client to validate nutanix resources
+type Validator struct {
+	v3Client client
+}
+
+// NewValidator returns a new validator client
+func NewValidator(config *anywherev1.NutanixDatacenterConfig, creds basicAuthCreds) (*Validator, error) {
+	url := fmt.Sprintf("%s:%d", config.Spec.Endpoint, config.Spec.Port)
+	nutanixCreds := prismgoclient.Credentials{
+		URL:      url,
+		Username: creds.username,
+		Password: creds.password,
+		Endpoint: config.Spec.Endpoint,
+	}
+	client, err := v3.NewV3Client(nutanixCreds)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Validator{v3Client: client.V3}, nil
+}
+
+// ValidateMachineConfig validates the Prism Element cluster, subnet, and image for the machine
+func (v *Validator) ValidateMachineConfig(ctx context.Context, config *anywherev1.NutanixMachineConfig) error {
+	var errors error
+	if err := v.validateClusterConfig(ctx, config.Spec.Cluster); err != nil {
+		errors = multierr.Append(errors, err)
+	}
+
+	if err := v.validateSubnetConfig(ctx, config.Spec.Subnet); err != nil {
+		errors = multierr.Append(errors, err)
+	}
+
+	if err := v.validateImageConfig(ctx, config.Spec.Image); err != nil {
+		errors = multierr.Append(errors, err)
+	}
+
+	return errors
+}
+
+func (v *Validator) validateClusterConfig(ctx context.Context, identifier anywherev1.NutanixResourceIdentifier) error {
+	switch identifier.Type {
+	case anywherev1.NutanixIdentifierName:
+		if identifier.Name == nil || *identifier.Name == "" {
+			return fmt.Errorf("missing cluster name")
+		} else {
+			clusterName := *identifier.Name
+			clusterUUID, err := findClusterUUIDByName(ctx, v.v3Client, clusterName)
+			if err != nil {
+				return fmt.Errorf("failed to find cluster with name %q: %v", clusterName, err)
+			}
+
+			identifier.Type = anywherev1.NutanixIdentifierUUID
+			identifier.UUID = clusterUUID
```

Comment:
Whats the intent with these lines? It seems like we're updating some configuration, but this is a validation function so I'm not sure why it would be mutating data.

This seems to be happening in several places below.

Reply:
Good catch; we don't need that here.